Sources/SwiftDiagnostics/SyntaxHighlightRangeCalculator.swift

```diff
@@ -0,0 +1,225 @@
+  ///
+  /// This struct is used to manage how leading and trailing trivia are treated for each highlight.
+  /// It stores the actual syntax highlight and two Boolean flags for trivia settings.
+  ///
+  /// - Important:
+  ///   - `afterLeadingTrivia`: When `true`, the starting column will be computed after considering leading trivia.
+  ///   - `afterTrailingTrivia`: When `false`, the ending column will be computed before considering trailing trivia.
+  private struct ExtendedHighlight {
+    /// The syntax element to be highlighted.
+    let highlight: Syntax
+    /// Flag to indicate whether leading trivia should be included in the highlight range.
+    var afterLeadingTrivia = true
+    /// Flag to indicate whether trailing trivia should be included in the highlight range.
+    var afterTrailingTrivia = false
+  }
+
+  /// Computes the ranges of highlightable regions on a specific line of source code.
+  ///
+  /// This method computes the start and end columns for each highlight in the line,
+  /// taking into account whether the leading or trailing trivia should be included for each highlight.
+  /// After computing these ranges, they are sorted based on their starting columns.
+  /// Finally, any overlapping ranges are merged into single contiguous ranges.
+  ///
+  /// - Parameters:
+  ///   - annotatedSourceLine: A tuple containing an array of Syntax highlights and the source string for the line.
+  ///   - lineNumber: The line number for which highlight ranges are to be computed.
+  ///   - tree: The Syntax tree containing the entire source code.
+  ///   - sourceLocationConverter: An instance for converting source locations.
+  ///
+  /// - Returns: An array of highlight ranges, represented by `Range<Int>` instances for the start and end columns.
+  ///
+  /// - Note:
+  ///   1. If the highlight is the first in the line, its leading trivia is always excluded.
+  ///   2. If a highlight's end aligns with the start of the next highlight, their shared trivia settings are adjusted.
+  ///   3. The last highlight in the line never includes trailing trivia.
+  @_spi(Testing) public static func computeHighlightRanges(
+    forLine annotatedSourceLine: (highlights: [Syntax], sourceString: String),
+    at lineNumber: Int,
+    fromTree tree: some SyntaxProtocol,
+    usingSourceLocationConverter sourceLocationConverter: SourceLocationConverter
+  ) -> [Range<Int>] {
+    // Initialize an array of extended highlights, based on provided Syntax highlights.
+    var extendedHighlights = annotatedSourceLine.highlights.map { ExtendedHighlight(highlight: $0) }
+
+    // Loop through each extended highlight to fine-tune its trivia settings.
+    for extendedHighlightsIndex in extendedHighlights.indices {
+      adjustTriviaSettingsForHighlight(
+        atIndex: extendedHighlightsIndex,
+        in: &extendedHighlights,
+        usingSourceLocationConverter: sourceLocationConverter
+      )
+    }
```

Comment:
I am not a fan of creating the `extendedHighlights` array and then fixing it up later. What I would do instead is:
Iterate over the highlighted nodes. If the node’s `positionBeforeTrailingTrivia` is equal to the last node’s `positionAfterTrailingTrivia`, extend the last highlight range in the array. If not, add a new entry to it.